47495: Update Enterprise Installing an Embedded Cluster

jonquil2002 opened the pull request:
SC: https://app.shortcut.com/replicated/story/47495/style-procedure-improvements-to-installing-with-the-kubernetes-installer

Preview: https://deploy-preview-381--replicated-docs.netlify.app/enterprise/installing-embedded-cluster

paigecalvert commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,124 @@
+## Prerequisites
 
-Before you install, ensure that you meet the system requirements. For more information, see [Kubernetes installer requirements](installing-embedded-cluster-requirements).
+Complete the following items before you perform this task:
```
s/ could remove the word "items"

paigecalvert commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,124 @@
+When installing on a highly available cluster, the script prompts for a load balancer address. In the absence of a load balancer, all traffic is routed to the first primary node.
 
-The health check for an apiserver is a TCP check on the port the kube-apiserver listens on (default value :6443).
+If you decide to use a load balancer, the load balancer should be:
```
should --> must ?

paigecalvert commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,124 @@
+To install an application and the admin console with high availability in an online environment, run:
 
 ```bash
 curl -sSL https://kurl.sh/APP-SLUG | sudo bash -s ha
````
An idea as I'm reading these procedures back to back: Since the commands for installing with HA are the same just with the added `-ha` flag, maybe these can be turned into optional steps in the main procedures.

For the optional load balancer part, maybe that could be added to the prerequisites section. For example, "If you are installing on a high availability cluster and want to use a load balancer, ensure the your load balancer meets these requirements..."

paigecalvert commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,124 @@
 ## Join Primary and Secondary Nodes
 
-Visit the `/cluster/manage` page in the admin console to generate scripts for joining additional secondary and primary nodes.
+You can generate commands in the admin console to join additional primary and secondary nodes.
```
Do you know what the context for this procedure is? It looks like the only other place we mention nodes on this page is where we say that if you don't use a load balancer, traffic goes to the primary node. Could be helpful to clarify the "why" on this section; I wasn't able to sort out how it fits in with installing.

(This could be a spin-off story if you think that's better. I think it would be addressing more of a content gap in terms of the "why")

paigecalvert commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,124 @@
+You can preconfigure the load balancer by passing in the `load-balancer-address=<host:port>` flag.
 
-In the absence of a load balancer, all traffic will be routed to the first primary.
+The health check for an apiserver is a TCP check on the port that the kube-apiserver listens on (the default value is `:6443`).
```
s/ The parenthetical phrase could be its own sentence

paigecalvert reviewed (approved):
Looks good @jonquil2002 ! Just small notes, and one suggestion around the separate HA procedures

paigecalvert reviewed (approved):
Looks great!

One thing I noticed in the air gap HA procedure: 

For air gap, I believe you'll still need to do the `curl` and `tar` commands before the `cat` install command. (I.e., the commands should look identical, with the only difference being that one of them has the -ha flag). The kurl docs have more info: https://kurl.sh/docs/install-with-kurl/#airgapped-usage

jonquil2002 commented:
@paigecalvert  - Yep, that's what I was thinking too and was planning to verify that. It just makes sense that they would, not sure why it wasn't documented originally.

jonquil2002 commented:
@JustinN-Replicated @ajp-io  Would one of you please do a technical review of this updated topic? In addition to heavy reformatting, I added a business reason for joining additional nodes and added the missing command from high availability air gap, so I want to make sure that it's technically correct. Thanks.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,124 @@
-cat install.sh | sudo bash -s airgap
-```
-Where `FILENAME` is the name of the kURL air gap `.tar.gz` file.
+Additionally, if you are installing with high availability and want to use a load balancer, ensure that your load balancer is:
````
Throughout this doc, I would refer to this as an external load balancer to distinguish it from the internal load balancer. Since they will almost certainly use a load balancer if they're installing in HA, but these requirements will only be pertinent if they bring their own, not if they use the internal one.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,124 @@
+Additionally, if you are installing with high availability and want to use a load balancer, ensure that your load balancer is:
+- A TCP forwarding load balancer
+- Configured to distribute traffic to all healthy control plane nodes in its target list
+- (Optional) Preconfigured by passing in the `load-balancer-address=<host:port>` flag
```
Oh great, you mention this here (I mentioned this flag in a comment above). Maybe we should similarly mention pre-configuring the internal load balancer.

jonquil2002 commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,124 @@
+- To install without high availability, run the installation command provided by the application vendor:
+
+  ```bash
+  curl -sSL https://kurl.sh/APP-SLUG | sudo bash
````
Thanks! This is an old tutorial, and I think there are many old kurl.sh references throughout the docs. I'll create an SC story to scrub the rest of the docs. But I fixed it for this tutorial.

jonquil2002 commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,124 @@
 
 :::note
-The air gap `.tar.gz` includes only the admin console components, which are required to install the application.
+With KOTS v1.67.0 and later, you can install a specific version of the application. Use the `app-version-label` flag and the version label for a particular version of your vendor's application. For example, `curl https://kurl.sh/supergoodtool | sudo bash -s app-version-label=3.1.0`.
```
Good call.

jonquil2002 commented:
@ajp-io  Hey, Alex. I made some substantial changes re: the load balancer info and also separated out the previous versions installation info.  I'll put a hold on merging this topic until the other version topic gets merged. Please review this again. Thanks.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,178 @@
+      ```
+      Replace `APP-SLUG` with the unique slug for the application. The application slug is included in the installation script provided by the vendor.
+
+    1. If you did not preconfigure a load balancer, you are prompted during the installation. Do one of the following:
+
+        - If you are using the internal load balancer, pass the following flag:
+
+          ```bash
+          ekco-enable-internal-load-balancer
+          ```
+
+        - If you are using an external load balancer, pass the following flag:
+
+          ```bash
+          load-balancer-address=HOST:PORT
+          ```
````
The flags would be passed to the install command above (e.g., `| sudo bash -s ha ekco-enable-internal-load-balancer`). That's the pre-configuring option. But if they don't preconfigure the load balancer (i.e., they don't pass one of those flags), then they are prompted for a load balancer address. Leaving it blank uses the internal load balancer, or you can pass the address of an external load balancer.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,178 @@
-cat install.sh | sudo bash -s airgap
-```
-Where `FILENAME` is the name of the kURL air gap `.tar.gz` file.
+1. Download and extract the kURL air gap `.tar.gz` file. The air gap `.tar.gz` includes only the admin console components, which are required to install the application.
````
I assume this is trying to say that the .tar.gz file doesn't include the application, but it feels a bit incorrect to say that it only includes the admin console components. It includes all the Kubernetes cluster components and the admin console. So if my assessment of what this is trying to say is correct, this might be more accurate.
```suggestion
1. Download and extract the kURL air gap `.tar.gz` file. The air gap `.tar.gz` file only includes the Kubernetes and admin console components. It doesn't include the application.
```

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,178 @@
-:::
+      ```bash
+      curl -LO https://k8s.kurl.sh/bundle/FILENAME.tar.gz
+      tar xvzf FILENAME.tar.gz
````
```suggestion
      tar -xvzf FILENAME.tar.gz
```

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,178 @@
-## Install with High Availability
+        ```bash
+        curl -LO https://k8s.kurl.sh/bundle/FILENAME.tar.gz
+        tar xvzf FILENAME.tar.gz
````
```suggestion
        tar -xvzf FILENAME.tar.gz
```

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,178 @@
+      1. If you did not preconfigure a load balancer, you are prompted during the installation. Do one of the following:
 
-The health check for an apiserver is a TCP check on the port the kube-apiserver listens on (default value :6443).
+          - If you are using the internal load balancer, pass the following flag:
```
Same comment here about the flags as above.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,178 @@
+          load-balancer-address=HOST:PORT
+          ```
+
+          Replace `HOST:PORT` with the host and port number of the external load balancer.
````
At the end, do we need to mention how to install the application? You would just open the admin console and upload the license, but since the air gap instructions mention the `kots install` command for installing the app, should we mention the app here too? Or is this only about installing the cluster and we shouldn't mention the app?

ajp-io replied:
Perfectly fine, thanks!

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,178 @@
 
-When installing on a highly available cluster, the script will prompt for a load balancer address.
-The load balancer can be preconfigured by passing in the `load-balancer-address=<host:port>` flag.
+        :::note
+        For both installation types, you can construct the URL for the air gap bundle by prefixing the URL path for online installations with `/bundle` and adding `.tar.gz` to the end. For more information, see [Install in an Online Environment](#install-in-an-online-environment).
+        :::
```
This note ends up nested under the "To install with high availability section", which to me makes it unclear that "both installation types" refers to HA and non-HA. It would make more sense if the note wasn't nested within one of those two types.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,178 @@
+
+To add primary and secondary nodes:
+
+1. (Air gap only) Download and extract the `.airgap` bundle on the remote node before running the join command.
```
I don't think this is actually referring to the application `.airgap` file. I think it is the `.tar.gz` file. The reason is, for air gap, you can't pull the join script from the internet (see how the command below is curling k8s.kurl.sh?). So instead, you have to have the `.tar.gz` file present on the remote node because it contains the `join.sh` script you're going to run.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,178 @@
+  primary-host=203.0.113.6
+  ```
+
+## Install Previous Versions in an Online Environment
````
I'm curious about this part here. This looks like it's more related to the other PR we have in progress about installing previous versions of the application. Are you also adding some context on that as part of this PR? I wasn't sure if this got mixed up with this one or if it's intentional.

jonquil2002 replied:
Intentional, as we discussed, and we'll postpone making these changes in a future PR. For now, I restored the notes about --app-version-label

Thank you so much for meeting with me.

ajp-io reviewed (changes requested):
One thing we might need a more technical SME for is, when should HA mode be used? Should I use it for single-node installations, or is it only relevant for multi-node installs? I also think there are rules about how many primary nodes you need (maybe it has to be an odd number), so that's also something it would be good to document here. Could be a separate iteration though once we nail this content. Up to you.

jonquil2002 commented:
@paigecalvert  @ajp-io  The combined HA and non-HA content got confusing to read after adding the load balancing info, so I separated them again. Please approve if possible. Thanks for all of your great feedback!

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,174 @@
+Additionally, if you are installing in high availability mode:
 
-For example:
+  - (Optional) If your vendor has enabled the internal load balancer, you can preconfigure it by passing `| sudo bash -s ha ekco-enable-internal-load-balancer`.
```
Small thing… The vendor doesn't have to enable it—the flag _is_ what enables it. So it's more like, "If you are going to use the internal load balancer…"

jonquil2002 replied:
@ajp-io Ok, oh, I thought that the vendor had to include the ECKO add-on and configure it in the manifest on their end. But it sounds like you're saying the internal load balancer is included by default for all Kubernetes installers?

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,174 @@
+  - For an external load balancer, ensure that your load balancer is:
+    - A TCP forwarding external load balancer.
+    - Configured to distribute traffic to all healthy control plane nodes in its target list.
+    - (Optional) Preconfigured with the load balancer address by passing the `load-balancer-address=HOST:PORT` flag.
```
If you combine the preamble before the colon and this bullet as is, it would read, "ensure that your load balancer is preconfigured with the load balancer address." But the LB doesn't have to be preconfigured with the LB address—the installer does. So I think this is clearer.
```suggestion
    - (Optional) Preconfigured by passing the `load-balancer-address=HOST:PORT` flag.
```

jonquil2002 replied:
I made your change anyway. My question is really to help me understand this feature better. Thanks!

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,174 @@
+
+To install the admin console and an application in an air gapped environment:
+
+1. Download and extract the kURL air gap `.tar.gz` file. The file only includes the Kubernetes and admin console components. It does not include the application.
+1. Run the following commands:
+
+    ```bash
+    curl -LO https://k8s.kurl.sh/bundle/FILENAME.tar.gz
+    tar -xvzf FILENAME.tar.gz
````
The `curl` command does the download and the `tar` command here does the extraction, so that is a bit repetitive since step 1 says to download and extract. Maybe these should just be one step, or however you want to resolve that.

ajp-io replied:
Down lower we just say, "Run the following commands," so we could do that again.

jonquil2002 commented:
@paigecalvert Ready for final review/approval when you get a chance. Thanks!

paigecalvert reviewed (approved):
Lgtm!

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,173 @@
+  ```bash
+  curl -sSL https://k8s.kurl.sh/APP_SLUG | sudo bash
+  ```
+  Replace `APP-SLUG` with the unique slug for the application. The application slug is included in the installation command provided by the vendor.
````
```suggestion
  Replace `APP_SLUG` with the unique slug for the application. The application slug is included in the installation command provided by the vendor.
```

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,173 @@
+    - If you are using an external load balancer, pass the load balancer address:
 
-To install with high availability in an online environment, run:
+      ```bash
+      load-balancer-address=HOST:PORT
+      ```
````
So to pass the load balancer address in this case, there is just a prompt that asks you to input the address. So you're not passing a flag like is indicated here. I think we can just remove these three lines (and the lines below about replacing stuff) and tell them to pass the LB address. Let me know if that makes sense. They should know how to do this when they see it—it's just a standard input.

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
````diff
@@ -1,100 +1,173 @@
+
+    - If you are using an external load balancer, pass the load balancer address:
+
+      ```bash
+      load-balancer-address=HOST:PORT
+      ```
+
+      Replace:
+
+      - `HOST` with the external load balancer host.
+      - `PORT` with the port number of the external load balancer.
````
Same as above. This stuff isn't really necessary because it's not a flag they're passing. They're literally just typing in the address at the prompt.

jonquil2002 commented:
@ajp-io Changes made - ready for approval?

ajp-io commented on docs/enterprise/installing-embedded-cluster.md:
```diff
@@ -1,100 +1,155 @@
+Additionally, if you are installing in high availability mode:
 
-For example:
+  - (Optional) If you are going to use the internal load balancer, you can preconfigure it by passing `| sudo bash -s ha ekco-enable-internal-load-balancer`.
```
Rereading it, I question this line since `| sudo bash` will already be there, and you really should just pass `-s ha ekco-enable-internal-load-balancer`, assuming they know how to do that. But that's not a dealbreaker, so I'm approving. Can handle separately if we want to.

ajp-io replied:
Just to clarify this, because I don't know if it's clear as I reread it: the default command includes the `| sudo bash` part, so it might make more sense to tell people to just pass `-s ha ekco-enable-internal-load-balancer`, since that's the only net new thing they need to add. But like I said before, I can live with it either way if we want to just merge this.

jonquil2002 replied:
Typically, we are supposed to document the entire command line entry, even if something is already there. So something like this would be correct formatting (even though this is not a full command, just showing this for demo purposes):

`$ curl install`

Not sure if this helps?

ajp-io replied:
If we're supposed to show the whole entry, then should we be showing all the way back to `curl` instead of starting in the middle with `| sudo bash`?

jonquil2002 replied:
Yep, I think so. I could open another PR to add the missing parts. Do you want to supply the full command here, s I'm not sure what it should be?
